debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java

```diff
@@ -297,34 +348,96 @@ protected abstract SchemaChangeEvent getCreateTableEvent(RelationalSnapshotConte
             }
+        }
 
-            LOGGER.debug("Snapshotting table {}", tableId);
+        int snapshotOrderByRowCount = connectorConfig.snapshotOrderByRowCount();
+        if (snapshotOrderByRowCount != 0) {
+            LOGGER.info("Sort tables by row count {}", snapshotOrderByRowCount > 0 ? "asc" : "desc");
+            rowCountTables = rowCountTables.entrySet().stream()
+                    .sorted(Map.Entry.comparingByValue((a, b) -> snapshotOrderByRowCount * ((Long) a.orElse(0)).compareTo(b.orElse(0))))
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e1, LinkedHashMap::new));
+        }
 
-            createDataEventsForTable(sourceContext, snapshotContext, snapshotReceiver, snapshotContext.tables.forTable(tableId), tableOrder++, tableCount);
+        Queue<O> offsets = new ConcurrentLinkedQueue<O>();
+        offsets.add(snapshotContext.offset);
+        for (int i = 1; i < snapshotMaxThreads; i++) {
+            offsets.add(copyOffset(snapshotContext));
+        }
+
+        try {
+            int tableCount = rowCountTables.size();
+            int tableOrder = 1;
+            for (TableId tableId : rowCountTables.keySet()) {
+                boolean firstTable = tableOrder == 1 && snapshotMaxThreads == 1;
+                boolean lastTable = tableOrder == tableCount && snapshotMaxThreads == 1;
+                String selectStatement = queryTables.get(tableId);
+                OptionalLong rowCount = rowCountTables.get(tableId);
+
+                Callable<Exception> callable = createDataEventsForTableCallable(sourceContext, snapshotContext, snapshotReceiver,
+                        snapshotContext.tables.forTable(tableId), firstTable, lastTable, tableOrder++, tableCount, selectStatement, rowCount, connectionPool, offsets);
+                completionService.submit(callable);
+            }
+
+            Exception firstError = null;
+            for (int i = 0; i < tableCount; i++) {
+                Exception error = completionService.take().get();
+                if (error != null) {
+                    if (firstError == null) {
+                        firstError = error;
+                    }
+                    else {
+                        LOGGER.debug("Snapshot task failed", error);
+                    }
+                }
+            }
+            if (firstError != null) {
+                throw firstError;
+            }
+        }
+        finally {
+            executorService.shutdownNow();
+            executorService.awaitTermination(10, TimeUnit.SECONDS);
         }
 
         releaseDataSnapshotLocks(snapshotContext);
-        snapshotContext.offset.preSnapshotCompletion();
+        for (O offset : offsets) {
+            offset.preSnapshotCompletion();
+        }
         snapshotReceiver.completeSnapshot();
-        snapshotContext.offset.postSnapshotCompletion();
+        for (O offset : offsets) {
+            offset.postSnapshotCompletion();
```

Comment:
The called method manipulates sourceInfo. Is not it lost in that case?

Reply:
@jpechane can you explain? what is supposed to be lost? each offset has its own sourceInfo.